File: Sample/EventStoreDB/Simple/ECommerce/ShoppingCarts/ProductItems/PricedProductItem.cs
using System;

namespace ECommerce.ShoppingCarts.ProductItems
{
    public record PricedProductItem(
        ProductItem ProductItem,
        decimal UnitPrice
    )
    {
        public Guid ProductId => ProductItem.ProductId;
        public int Quantity => ProductItem.Quantity;

        public decimal TotalPrice => Quantity * UnitPrice;

        public static PricedProductItem From(ProductItem productItem, decimal? unitPrice)
        {
            return unitPrice switch
            {
                null => throw new ArgumentNullException(nameof(unitPrice)),
                <= 0 => throw new ArgumentOutOfRangeException(nameof(unitPrice),
                    "Unit price has to be positive number"),
                _ => new PricedProductItem(productItem, unitPrice.Value)
            };
        }

        public bool MatchesProductAndUnitPrice(PricedProductItem pricedProductItem)
        {
            return ProductId == pricedProductItem.ProductId && UnitPrice == pricedProductItem.UnitPrice;
        }

        public PricedProductItem MergeWith(PricedProductItem productItem)
        {
            if(ProductId != productItem.ProductId)
                throw new ArgumentException("Product ids do not match.");
            if(UnitPrice != productItem.UnitPrice)
                throw new ArgumentException("Product unit prices do not match.");

            return new PricedProductItem(
                new ProductItem(ProductId, productItem.Quantity + productItem.Quantity),
                UnitPrice
            );
        }
    }
}

File: Sample/ECommerce.Equinox/ECommerce.Domain/PricedProductItem.fs
namespace ECommerce.Domain

open System

type ProductItem = { productId : ProductId; quantity : int }

type PricedProductItem = { productItem : ProductItem; unitPrice : decimal } with

    member x.ProductId = x.productItem.productId
    member x.Quantity = x.productItem.quantity
    member x.TotalPrice = decimal x.Quantity * x.unitPrice

    static member From(productItem, ?unitPrice : decimal) =
        match unitPrice with
        | None -> nullArg (nameof(unitPrice))
        | Some price when price <= 0m -> raise <| ArgumentOutOfRangeException(nameof unitPrice, "Unit price has to be positive number")
        | Some price -> { productItem = productItem; unitPrice = price }

    member x.MatchesProductAndUnitPrice(pricedProductItem : PricedProductItem) =
        x.ProductId = pricedProductItem.ProductId && x.unitPrice = pricedProductItem.unitPrice

    member x.MergeWith(productItem : PricedProductItem) =
        if x.ProductId <> productItem.ProductId then raise <| ArgumentException "Product ids do not match."
        if x.unitPrice <> productItem.unitPrice then raise <| ArgumentException "Product unit prices do not match."
        // TODO fix bug in source: new ProductItem(ProductId, productItem.Quantity + productItem.Quantity),
        { productItem = { productId = x.ProductId; quantity = x.Quantity + productItem.Quantity }; unitPrice = x.unitPrice }
